File: pkg/models/investment.go
package models

// InvestmentPosition binds a quantity ledger to one monetary cost account.
// Investment tables are additive; legacy Account.Balance remains minor currency units.
type InvestmentPosition struct {
	Id              string `xorm:"VARCHAR(36) PK" json:"id"`
	Uid             int64  `xorm:"INDEX NOT NULL" json:"-"`
	Name            string `xorm:"VARCHAR(128) NOT NULL" json:"name"`
	AssetType       string `xorm:"VARCHAR(16) NOT NULL" json:"assetType"`
	Unit            string `xorm:"VARCHAR(16) NOT NULL" json:"unit"`
	Currency        string `xorm:"VARCHAR(3) NOT NULL" json:"currency"`
	CostAccountId   int64  `xorm:"UNIQUE NOT NULL" json:"costAccountId,string"`
	Quantity        string `xorm:"VARCHAR(40) NOT NULL" json:"quantity"`
	Cost            int64  `xorm:"NOT NULL" json:"cost,string"`
	Realized        int64  `xorm:"NOT NULL" json:"realized,string"`
	Algorithm       string `xorm:"VARCHAR(16) NOT NULL" json:"algorithm"`
	Version         int64  `xorm:"NOT NULL" json:"version"`
	CreatedUnixTime int64  `xorm:"NOT NULL" json:"createdAt"`
}

// InvestmentRevision is append-only. Current facts are the greatest Revision per OperationId.
type InvestmentRevision struct {
	Id                 string `xorm:"VARCHAR(36) PK" json:"id"`
	Uid                int64  `xorm:"INDEX NOT NULL" json:"-"`
	PositionId         string `xorm:"VARCHAR(36) INDEX NOT NULL" json:"positionId"`
	OperationId        string `xorm:"VARCHAR(36) UNIQUE(inv_revision) NOT NULL" json:"operationId"`
	Revision           int64  `xorm:"UNIQUE(inv_revision) NOT NULL" json:"revision"`
	Sequence           int64  `xorm:"NOT NULL" json:"sequence"`
	Kind               string `xorm:"VARCHAR(16) NOT NULL" json:"kind"`
	OccurredAt         int64  `xorm:"NOT NULL" json:"occurredAt"`
	TimezoneUtcOffset  int16  `xorm:"NOT NULL" json:"utcOffset"`
	Quantity           string `xorm:"VARCHAR(40) NOT NULL" json:"quantity"`
	Gross              int64  `xorm:"NOT NULL" json:"gross,string"`
	Fee                int64  `xorm:"NOT NULL" json:"fee,string"`
	CashAccountId      int64  `xorm:"NOT NULL" json:"cashAccountId,string"`
	TransferCategoryId int64  `xorm:"NOT NULL" json:"transferCategoryId,string"`
	IncomeCategoryId   int64  `xorm:"NOT NULL" json:"incomeCategoryId,string"`
	ExpenseCategoryId  int64  `xorm:"NOT NULL" json:"expenseCategoryId,string"`
	Cancelled          bool   `xorm:"NOT NULL" json:"cancelled"`
	Comment            string `xorm:"VARCHAR(255)" json:"comment"`
	RecordedAt         int64  `xorm:"NOT NULL" json:"recordedAt"`
}
type InvestmentPosting struct {
	Id                   string `xorm:"VARCHAR(36) PK"`
	Uid                  int64  `xorm:"INDEX NOT NULL"`
	PositionId           string `xorm:"VARCHAR(36) INDEX NOT NULL"`
	OperationId          string `xorm:"VARCHAR(36) NOT NULL"`
	RevisionId           string `xorm:"VARCHAR(36) NOT NULL"`
	TransactionId        int64  `xorm:"UNIQUE NOT NULL"`
	RelatedTransactionId int64  `xorm:"INDEX NOT NULL"`
	PositionVersion      int64  `xorm:"NOT NULL"`
	Active               bool   `xorm:"NOT NULL"`
}
type InvestmentCommand struct {
	Id          string `xorm:"VARCHAR(36) PK"`
	Uid         int64  `xorm:"UNIQUE(inv_command) NOT NULL"`
	RequestKey  string `xorm:"VARCHAR(64) UNIQUE(inv_command) NOT NULL"`
	PayloadHash string `xorm:"VARCHAR(64) NOT NULL"`
	Result      string `xorm:"TEXT NOT NULL"`
}

// InvestmentAttachment owns a receipt independently of regenerated cash postings.
type InvestmentAttachment struct {
	PictureId   int64  `xorm:"PK" json:"pictureId,string"`
	Uid         int64  `xorm:"INDEX NOT NULL" json:"-"`
	PositionId  string `xorm:"VARCHAR(36) INDEX NOT NULL" json:"positionId"`
	OperationId string `xorm:"VARCHAR(36) NOT NULL" json:"operationId"`
	RevisionId  string `xorm:"VARCHAR(36) NOT NULL" json:"revisionId"`
	Extension   string `xorm:"VARCHAR(10) NOT NULL" json:"extension"`
	RecordedAt  int64  `xorm:"NOT NULL" json:"recordedAt"`
}

// InvestmentValuationSetting changes display estimates only, never the cost ledger.
type InvestmentValuationSetting struct {
	PositionId    string `xorm:"VARCHAR(36) PK" json:"positionId"`
	Uid           int64  `xorm:"INDEX NOT NULL" json:"-"`
	Version       int64  `xorm:"NOT NULL" json:"version"`
	Mode          string `xorm:"VARCHAR(16) NOT NULL" json:"mode"`
	ManualPrice   string `xorm:"VARCHAR(40)" json:"manualPrice"`
	ManualAsOf    int64  `json:"manualAsOf"`
	MaxAgeMinutes int64  `json:"maxAgeMinutes"`
	UpdatedAt     int64  `json:"updatedAt"`
}

File: src/models/investment.ts
export interface InvestmentPosition {
    definitionId?: string; unitName?: string; quantityPrecision?: number;
    id: string; name: string; assetType: string; unit: string; currency: string;
    costAccountId: string; quantity: string; cost: string; realized: string;
    algorithm: string; version: number;
}
export interface InvestmentOperation {
    operationId: string; revision: number; kind: 'opening' | 'buy' | 'sell';
    occurredAt: number; utcOffset: number; quantity: string; gross: string; fee: string;
    cashAccountId: string; transferCategoryId: string; incomeCategoryId: string;
    expenseCategoryId: string; cancelled: boolean; comment: string;
}
export interface InvestmentRequest extends Omit<InvestmentOperation, 'revision'> {
    entryPurpose?: 'asset_purchase'; assetDefinitionId?:string;
    requestKey: string; positionId: string; expectedVersion: number;
}
export interface InvestmentCalculation {
    quantity: string; cost: string; realized: string;
    effects: { eventId: string; cashDelta: string; allocatedCost: string; realized: string }[];
}
export interface InvestmentDetail {
    history?: (InvestmentOperation & { id: string; recordedAt: number })[];
    attachments: { pictureId: string; operationId: string; extension: string }[] | null;
    position: InvestmentPosition; operations: InvestmentOperation[]; calculation: InvestmentCalculation;
}
// Currency entry is parsed exactly. No binary floating point multiplication.
export function investmentMinorUnits(value: string): string {
    if (!/^(0|[1-9]\d*)(\.\d{1,2})?$/.test(value)) throw new Error('请输入最多两位小数的非负金额');
    const [whole, fraction = ''] = value.split('.');
    const amount = BigInt(whole!) * 100n + BigInt(fraction.padEnd(2, '0'));
    if (amount > 9999999999999n) throw new Error('金额超出支持范围');
    return amount.toString();
}
export function investmentMoney(minor: string): string {
    const value = BigInt(minor); const absolute = value < 0n ? -value : value;
    return `${value < 0n ? '-' : ''}${absolute / 100n}.${(absolute % 100n).toString().padStart(2, '0')}`;
}

export function investmentAverage(cost: string, quantity: string): string {
    if (!/^(0|[1-9]\d*)(\.\d{1,12})?$/.test(quantity)) return '—';
    const [whole, fraction = ''] = quantity.split('.');
    const q = BigInt(whole!) * 1000000000000n + BigInt(fraction.padEnd(12, '0'));
    if (q === 0n) return '—';
    const scaled = (BigInt(cost) * 100000000000000n + q / 2n) / q;
    return `${scaled / 10000n}.${(scaled % 10000n).toString().padStart(4, '0')}`;
}

export interface InvestmentDefinition { id:string; name:string; kind:string; unit:string; unitName:string; precision:number; version:number; inUse:boolean; }
